File: ios/ZoomablePdfScrollViewManager.m
#import <React/RCTViewManager.h>

@interface RCT_EXTERN_REMAP_MODULE(RNZoomablePdfScrollView, ZoomablePdfScrollViewManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(source, NSString)
RCT_EXPORT_VIEW_PROPERTY(annotations, NSString)
RCT_EXPORT_VIEW_PROPERTY(minZoom, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(maxZoom, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(edgeTapZone, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(pdfPaddingTop, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(pdfPaddingBottom, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(pdfBackgroundColor, UIColor)

// Drawing props
RCT_EXPORT_VIEW_PROPERTY(drawingMode, NSString)
RCT_EXPORT_VIEW_PROPERTY(strokeColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(strokeWidth, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(strokeOpacity, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(strokes, NSString)

// Events
RCT_EXPORT_VIEW_PROPERTY(onPdfError, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPdfLoadComplete, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPageChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onZoomChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onTap, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMiddleClick, RCTDirectEventBlock)

// Drawing events
RCT_EXPORT_VIEW_PROPERTY(onDrawingStart, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onDrawingEnd, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onUndoStateChange, RCTDirectEventBlock)

// Text annotation props
RCT_EXPORT_VIEW_PROPERTY(textColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(textFontSize, CGFloat)

// Commands
RCT_EXTERN_METHOD(undo:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(redo:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(resetZoom:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(scrollToPage:(nonnull NSNumber *)node page:(int)page animated:(BOOL)animated)
RCT_EXTERN_METHOD(clearStrokes:(nonnull NSNumber *)node page:(int)page)
RCT_EXTERN_METHOD(getAnnotations:(nonnull NSNumber *)node
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios/ZoomablePdfScrollViewManager.swift
@objc(ZoomablePdfScrollViewManager)
class ZoomablePdfScrollViewManager: RCTViewManager {

    @objc
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }

    override func view() -> UIView! {
        return ZoomablePdfScrollView()
    }

    @objc func resetZoom(_ node: NSNumber) {
        withPdfView(node) { $0.resetZoom() }
    }

    @objc func scrollToPage(_ node: NSNumber, page: Int, animated: Bool) {
        withPdfView(node) { $0.scrollToPage(page, animated: animated) }
    }

    @objc func undo(_ node: NSNumber) {
        withPdfView(node) { $0.undo() }
    }

    @objc func redo(_ node: NSNumber) {
        withPdfView(node) { $0.redo() }
    }

    @objc func clearStrokes(_ node: NSNumber, page: Int) {
        withPdfView(node) { $0.clearStrokes(page: page) }
    }

    @objc func getAnnotations(_ node: NSNumber,
                               resolver: @escaping RCTPromiseResolveBlock,
                               rejecter: @escaping RCTPromiseRejectBlock) {
        withPdfView(node, rejecter: rejecter) { resolver($0.getAnnotations()) }
    }

    private func withPdfView(_ node: NSNumber, rejecter: RCTPromiseRejectBlock? = nil, action: @escaping (ZoomablePdfScrollView) -> Void) {
        guard let uiManager = bridge.uiManager else {
            rejecter?("ERROR", "UIManager not available", nil)
            return
        }
        uiManager.addUIBlock { (_, viewRegistry) in
            guard let viewRegistry,
                  let view = viewRegistry[node] as? ZoomablePdfScrollView else {
                rejecter?("ERROR", "View not found", nil)
                return
            }
            action(view)
        }
    }
}
